### DISCUSSION TITLE
Convert MIW application into multi module gradle project

### DISCUSSION THREAD

**Discussion by nitin-vavdiya** [eclipse-tractusx/managed-identity-wallet#254] [Ideas]
Hey MIW Contributors,

The current MIW application is structured as a single-module Gradle project, which presents limitations in sharing common components such as DTOs, constants, interfaces, and utility classes as JAR files. Consequently, other projects like EDC and Revocation cannot easily utilize these shared components, leading to code duplication across projects.

Addressing this issue would yield several benefits:

1. It enables the definition of interfaces for storing keys in various storage systems, allowing customers to implement these interfaces based on their preferred storage provider (e.g., database, S3, Azure Key Vault, or HashiCorp Vault).
2. It facilitates the definition of interfaces for credential revocations or suspensions.

To foster collaboration and reduce redundancy in code, it's essential to restructure the MIW application to support the seamless sharing of common components among different projects. This can greatly enhance efficiency and maintainability across the board.

Project structure something like this:

![image](https://github.com/eclipse-tractusx/managed-identity-wallet/assets/45592624/984b0159-ae3f-4316-b5e8-3ebc99d8abff)

Ref: https://docs.gradle.org/current/userguide/multi_project_builds.html

**Comment by hkny** [comment]
Would this also simplify splitting the MIW into separate microservices? Issuer Service, Holder Wallet, Verification Service etc.?

**Comment by borisrizov-zf** [comment]
Thank you for your suggestion. I can see this is a very early draft. We'll need to do a very extensive analysis on the benefit and drawbacks of using this code structure, compared to the current monolithic approach.
I currently don't think this is viable, as the project still has a very unstable API. At least in the sense that new features are coming in, and already existing features are being modified. The CI is barely stable as-is, and this would require us to overhaul it completely. There is a lot of work to be done to get to this multi-module project and the current timeline doesn't seem to support this refactor at all

This would probably be best addressed in the post-consortia time.
